pallets/creditcoin/src/ocw/collect_coins.rs

```diff
@@ -0,0 +1,757 @@
+use super::{
+	errors::{VerificationFailureCause, VerificationResult},
+	rpc::{self, EthBlock, EthTransaction, EthTransactionReceipt},
+	OffchainResult, ETH_CONFIRMATIONS,
+};
+
+use crate::pallet::{Config, Pallet};
+use crate::{
+	types::{AddressId, Blockchain, CollectCoins, UnverifiedCollectCoins},
+	ExternalAddress, ExternalAmount,
+};
+use sp_runtime::traits::UniqueSaturatedFrom;
+
+use ethabi::{Function, Param, ParamType, StateMutability, Token};
+use ethereum_types::{H160, U64};
+use frame_support::ensure;
+use sp_std::prelude::*;
+
+pub(crate) const CONTRACT_CHAIN: Blockchain = Blockchain::Ethereum;
+
+///exchange has been deprecated, use burn instead
+fn burn_vested_cc_abi() -> Function {
+	#[allow(deprecated)]
+	Function {
+		name: "exchange".into(),
+		inputs: vec![Param {
+			name: "value".into(),
+			kind: ParamType::Uint(256),
+			internal_type: None,
+		}],
+		outputs: vec![Param { name: "success".into(), kind: ParamType::Bool, internal_type: None }],
+		constant: false,
+		state_mutability: StateMutability::NonPayable,
+	}
+}
+pub fn validate_collect_coins(
+	to: &ExternalAddress,
+	amount: &mut Option<ExternalAmount>,
+	receipt: &EthTransactionReceipt,
+	transaction: &EthTransaction,
+	eth_tip: U64,
+) -> OffchainResult<()> {
+	ensure!(receipt.is_success(), VerificationFailureCause::TaskFailed);
+	let vesting_contract = hex::decode(b"a3EE21C306A700E682AbCdfe9BaA6A08F3820419").unwrap();
+	let vesting_contract = &H160::from(<[u8; 20]>::try_from(vesting_contract.as_slice()).unwrap());
+
+	let block_number = transaction.block_number.ok_or(VerificationFailureCause::TaskPending)?;
+
+	let diff = (eth_tip).checked_sub(block_number);
+	ensure!(diff.is_some(), VerificationFailureCause::TaskInFuture);
+
+	ensure!(diff.unwrap().as_u64() >= ETH_CONFIRMATIONS, VerificationFailureCause::TaskUnconfirmed);
+
+	if let Some(to) = &transaction.to {
+		ensure!(to == vesting_contract, VerificationFailureCause::IncorrectContract);
+	} else {
+		return Err(VerificationFailureCause::MissingReceiver.into());
+	}
+
+	if let Some(from) = &transaction.from {
+		ensure!(from[..] == to[..], VerificationFailureCause::IncorrectSender)
+	} else {
+		return Err(VerificationFailureCause::MissingSender.into());
+	}
+
+	let transfer_fn = burn_vested_cc_abi();
+	//is ignoring the selector a good idea? test? Same input, diff call (not exchange)?
+	ensure!(transaction.input.0.len() > 4, VerificationFailureCause::EmptyInput);
+	let inputs = transfer_fn.decode_input(&transaction.input.0[4..]).map_err(|e| {
+		log::error!("failed to decode inputs: {:?}", e);
+		VerificationFailureCause::AbiMismatch
+	})?;
+
+	let input_amount = match inputs.get(0) {
+		Some(Token::Uint(value)) => ExternalAmount::from(value),
+		_ => return Err(VerificationFailureCause::IncorrectInputType.into()),
+	};
+
+	*amount = Some(input_amount);
+
+	Ok(())
+}
+
+impl<T: Config> Pallet<T> {
+	///Amount is saturated to u128, don't exchange more than u128::MAX at once.
+	pub fn verify_cc_ocw(
```

Comment:
name should be `verify_collect_coins_ocw` for clarity.

Reply:
Will fix.